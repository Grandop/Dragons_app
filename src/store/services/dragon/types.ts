import { Dragon } from "../../../entities/dragon";

export type CreateDragonParams = Omit<Dragon, "id" | "createdAt">;

export type EditDragonParams = Omit<Dragon, "createdAt">;

export type DragonId = Pick<Dragon, "id">;
