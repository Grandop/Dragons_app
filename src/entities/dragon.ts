export interface Dragon {
  id: string;
  name: string;
  imageUrl: string;
  createdAt: string;
  type: string;
  histories: string[];
}
