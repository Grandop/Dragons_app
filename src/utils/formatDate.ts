import { format } from "date-fns";

export function formatDate(date: string): string {
  const parsedDate = new Date(date);
  const formattedDate = format(parsedDate, "dd/MM/yyyy");
  const hours = parsedDate.getHours();
  const minutes = parsedDate.getMinutes();
  return `${formattedDate} as ${hours}h:${minutes}m`;
}
