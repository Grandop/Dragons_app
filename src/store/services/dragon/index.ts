import type { CreateDragonParams, DragonId, EditDragonParams } from "./types";
import { Dragon } from "@entities/dragon";
import { createApi } from "@reduxjs/toolkit/query/react";
import { baseQuery } from "../../config/base-query";

export const DragonService = createApi({
  reducerPath: "dragon-service",
  baseQuery,
  tagTypes: ["Dragon"],
  endpoints: (builder) => ({
    getDragon: builder.query<Dragon[], void>({
      query: () => ({
        method: "GET",
        url: "/api/v1/dragon"
      }),
      providesTags: ["Dragon"]
    }),
    getDragonById: builder.query<Dragon, DragonId>({
      query: ({ id }: DragonId) => ({
        method: "GET",
        url: `/api/v1/dragon/${id}`
      }),
      providesTags: ["Dragon"]
    }),
    createDragon: builder.mutation<void, CreateDragonParams>({
      query: (data: CreateDragonParams) => ({
        method: "POST",
        url: "/api/v1/dragon",
        body: data
      }),
      invalidatesTags: ["Dragon"]
    }),
    editDragon: builder.mutation<void, EditDragonParams>({
      query: (data: EditDragonParams) => ({
        method: "PUT",
        url: `/api/v1/dragon/${data.id}`,
        body: data
      }),
      invalidatesTags: ["Dragon"]
    }),
    deleteDragon: builder.mutation<void, DragonId>({
      query: ({ id }: DragonId) => ({
        method: "DELETE",
        url: `/api/v1/dragon/${id}`
      }),
      invalidatesTags: ["Dragon"]
    })
  })
});

export const {
  useGetDragonQuery,
  useGetDragonByIdQuery,
  useCreateDragonMutation,
  useEditDragonMutation,
  useDeleteDragonMutation
} = DragonService;
