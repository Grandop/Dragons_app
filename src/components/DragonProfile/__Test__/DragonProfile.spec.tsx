import { Dragon } from "@entities/dragon";
import { render, screen } from "@testing-library/react";
import { theme } from "@theme";
import { ThemeProvider } from "styled-components";
import { describe, expect, it } from "vitest";
import { formatDate } from "@utils/formatDate";
import { DragonProfile } from "../index";

const renderWithTheme = (ui: React.ReactElement) =>
  render(<ThemeProvider theme={theme}>{ui}</ThemeProvider>);

describe("DragonProfile component", () => {
  const mockDragon: Dragon = {
    id: "1",
    name: "Test Dragon",
    type: "Fire",
    createdAt: "2024-01-01T00:00:00Z",
    histories: "This is a fierce dragon",
    imageUrl: "https://fakeimg.com/dragon.png"
  };

  it("should render the dragon name", () => {
    renderWithTheme(<DragonProfile dragon={mockDragon} />);
    expect(screen.getByText("Test Dragon")).toBeInTheDocument();
  });

  it("should render the dragon type", () => {
    renderWithTheme(<DragonProfile dragon={mockDragon} />);
    expect(screen.getByText("Fire")).toBeInTheDocument();
  });

  it("should render the createdAt formatted", () => {
    renderWithTheme(<DragonProfile dragon={mockDragon} />);
    const formattedDate = formatDate(mockDragon.createdAt);
    expect(screen.getByText(`Criado em ${formattedDate}`)).toBeInTheDocument();
  });

  it("should render the dragon history", () => {
    renderWithTheme(<DragonProfile dragon={mockDragon} />);
    expect(screen.getByText(/This is a fierce dragon/i)).toBeInTheDocument();
  });

  it("should render fallback if createdAt is missing", () => {
    renderWithTheme(
      <DragonProfile dragon={{ ...mockDragon, createdAt: "" }} />
    );
    expect(screen.getByText(/Data inválida/i)).toBeInTheDocument();
  });

  it("should render fallback if history is empty", () => {
    renderWithTheme(
      <DragonProfile dragon={{ ...mockDragon, histories: "" }} />
    );
    expect(
      screen.getByText(/Nenhuma descrição encontrada/i)
    ).toBeInTheDocument();
  });

  it("should match snapshot", () => {
    const { container } = renderWithTheme(
      <DragonProfile dragon={mockDragon} />
    );
    expect(container).toMatchSnapshot();
  });
});
