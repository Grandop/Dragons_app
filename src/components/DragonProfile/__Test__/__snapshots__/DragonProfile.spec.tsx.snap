// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DragonProfile component > should match snapshot 1`] = `
<div>
  <div
    class="sc-hvigdm ckrhaK"
  >
    <div
      class="sc-fhHczv ckJNIP"
    >
      <div
        class="sc-ggWZvA jRUOUy"
      >
        <div
          class="sc-kcLKEh cpQqLO"
          data-testid="avatar-container"
        >
          <img
            alt="Avatar"
            class="sc-dntSTA ilEuCB"
            src="https://fakeimg.com/dragon.png"
          />
        </div>
        <div
          class="sc-hjsuWn dcOXKT"
        >
          <h1
            class="sc-bRKDuR eoLTkw sc-jJLAfE dcqtOp"
            font-size="heading1"
          >
            Test Dragon
          </h1>
          <h3
            class="sc-bRKDuR fjtKsL sc-hwkwBN bVZfqO"
            font-size="heading3"
          >
            Criado em
             
            31/12/2023
          </h3>
          <div
            class="sc-kNOymR eytJCS"
          >
            Fire
          </div>
        </div>
      </div>
      <div
        class="sc-dYwGCk bLwZUA"
      >
        <h1
          class="sc-bRKDuR eoLTkw sc-lgpSej dIyoxh"
          font-size="heading1"
        >
          História ou Descrição
        </h1>
        <h3
          class="sc-bRKDuR fjtKsL sc-eqYatC ihZcnu"
          font-size="heading3"
        >
          This is a fierce dragon
        </h3>
      </div>
    </div>
  </div>
</div>
`;
